lib/frrscript.h

```diff
@@ -96,21 +97,88 @@ void frrscript_register_type_codecs(struct frrscript_codec *codecs);
+
+#define DECODE_ARGS(name, value)                                               \
+	do {                                                                   \
+		lua_getglobal(L, name);                                        \
+		DECODE_ARGS_WITH_STATE(L, value);                              \
+	} while (0)
+
+/*
+ * Maps the type of value to its encoder/decoder.
+ * Add new mappings here.
+ *
+ * L
+ *    Lua state
+ * scriptdir
+ *    Directory in which to look for scripts
+ */
+#define ENCODE_ARGS_WITH_STATE(L, value)                                       \
+	_Generic((value), \
+long long * : lua_pushintegerp,                                 \
+struct prefix * : lua_pushprefix,                               \
+struct interface * : lua_pushinterface,                         \
+struct in_addr * : lua_pushinaddr,                              \
+struct in6_addr * : lua_pushin6addr,                            \
+union sockunion * : lua_pushsockunion,                          \
+time_t * : lua_pushtimet,                                       \
+char * : lua_pushstring_wrapper,                                \
+struct attr * : lua_pushattr,                                   \
+struct peer * : lua_pushpeer,                                   \
+const struct prefix * : lua_pushprefix                          \
+)(L, value)
+
+#define DECODE_ARGS_WITH_STATE(L, value)                                       \
+	_Generic((value), \
+long long * : lua_decode_integerp,                              \
+struct prefix * : lua_decode_prefix,                            \
+struct interface * : lua_decode_interface,                      \
+struct in_addr * : lua_decode_inaddr,                           \
+struct in6_addr * : lua_decode_in6addr,                         \
+union sockunion * : lua_decode_sockunion,                       \
+time_t * : lua_decode_timet,                                    \
+char * : lua_decode_stringp,                                    \
+struct attr * : lua_decode_attr,                                \
+struct peer * : lua_decode_noop,                                \
+const struct prefix * : lua_decode_noop                         \
+)(L, -1, value)
 
 /*
  * Call script.
  *
  * fs
  *    The script to call; this is obtained from frrscript_load().
  *
- * env
- *    The script's environment. Specify this as an array of frrscript_env.
- *
  * Returns:
  *    0 if the script ran successfully, nonzero otherwise.
  */
-int frrscript_call(struct frrscript *fs, struct frrscript_env *env);
+int _frrscript_call(struct frrscript *fs);
 
+/*
+ * Wrapper for call script. Maps values passed in to their encoder
+ * and decoder types.
+ *
+ * fs
+ *    The script to call; this is obtained from frrscript_load().
+ *
+ * Returns:
+ *    0 if the script ran successfully, nonzero otherwise.
+ */
+#define frrscript_call(fs, ...)                                                \
+	({                                                                     \
```

Comment:
This should be `do { ... } while (0)` - that is the pattern we use for blocks everywhere else

Reply:
I wanted `frrscript_call()` to evaluate to 0/1 if success/failure, so I used a [statement expression](https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html) that evaluates to the last subexpression i.e. `ret`. instead of `do while`. So we can use it like `res = frrscript_call( ... )`. It's hard to get this behavior with macros wrapped in `do while` since the entire block evaluates to void.  
Some other macros defined with statement expressions are [here](https://github.com/dlqs/frr/blob/b16a58fc77748441e1175374b6fd875ec641e630/lib/compiler.h#L307)

Reply:
Then sounds good